google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java

```diff
@@ -62,22 +63,84 @@ public Batch newBatch() {
     return new BatchImpl(this);
   }
 
+  @Override
+  public Transaction newTransaction(TransactionOptions transactionOptions) {
+    return new TransactionImpl(this, transactionOptions);
+  }
+
   @Override
   public Transaction newTransaction() {
     return new TransactionImpl(this);
   }
 
+  static class ReadWriteTransactionCallable<T> implements Datastore.ReadWriteTransactionCallable<T>, Callable<T> {
+    private final Datastore datastore;
+    private final TransactionCallable<T> callable;
+    private TransactionOptions options;
+    private Transaction transaction;
+
+    ReadWriteTransactionCallable(Datastore datastore, TransactionCallable<T> callable, TransactionOptions options) {
+      this.datastore = datastore;
+      this.callable = callable;
+      this.options = options;
+      this.transaction = null;
+    }
+
+    Datastore getDatastore() {
+      return datastore;
+    }
+
+    TransactionOptions getOptions() {
+      return options;
+    }
+
+    Transaction getTransaction() {
+      return transaction;
+    }
+
+    @Override
+    public T run() {
+      transaction = datastore.newTransaction(options);
+      try {
+        T value = callable.run(transaction);
+        transaction.commit();
+        return value;
+      } catch (Exception ex) {
+        transaction.rollback();
+        throw DatastoreException.propagateUserException(ex);
+      } finally {
+        if (transaction.isActive()) {
+          transaction.rollback();
+        }
+        if (options.getModeCase().equals(TransactionOptions.ModeCase.READ_WRITE)) {
+          setPrevTransactionId(transaction);
+        }
+      }
+    }
+
+    void setPrevTransactionId(Transaction transaction) {
+      options =
+          options
+              .toBuilder()
+              .setReadWrite(
```

Comment:
Please format this  somewhat prettier (one option to consider is to create local variables, for example for the value passed to `.setReadWrite()`).

Reply:
Done.